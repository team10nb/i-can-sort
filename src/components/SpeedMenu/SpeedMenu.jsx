/*
    Author: Shiliang Chen, Ruizi Han

    The menu used to change speed.
    material-ui library is used for UI.
*/

import React from "react";
import Button from "@material-ui/core/Button";
import Menu from "@material-ui/core/Menu";
import MuiMenuItem from "@material-ui/core/MenuItem";
import { makeStyles, withStyles } from "@material-ui/core/styles";
import Tooltip from "@material-ui/core/Tooltip";
import Zoom from "@material-ui/core/Zoom";

const MenuItem = withStyles({
    root: {
        justifyContent: "center",
        fontSize: "10px",
        fontFamily: "inherit",
        fontWeight: "600",
    },
})(MuiMenuItem);

export default function SpeedMenu(props) {
    const { handleClick, handleClose, anchorEl, speed } = props;

    const useStyles = makeStyles({
        root: {
            display: "inline-block",
            width: "10px",
        },
        menu: {
            width: "80px",
        },
        button: {
            width: "10px",
            paddingLeft: "0px",
            paddingRight: "0px",
            fontSize: "12px",
            fontWeight: "600",
            fontFamily: "inherit",
        },
        customTooltip: {
            // I used the rgba color for the standard "secondary" color
            fontFamily: "inherit",
            fontSize: "1.2em",
            fontWeight: "700",
            paddingTop: "5px",
            paddingBottom: "6px",
            letterSpacing: "1px",
            marginTop: "3px"
        },
    });
    const classes = useStyles();

    return (
        <div className={classes.root}>
            <Tooltip
                title='Speed'
                TransitionComponent={Zoom}
                enterDelay={500}
                leaveDelay={200}
                classes={{ tooltip: classes.customTooltip }}
                arrow
            >
                <Button
                    aria-controls='simple-menu'
                    aria-haspopup='true'
                    onClick={handleClick}
                    className={classes.button}
                    style={{ minWidth: "40px" }}
                >
                    {speed}
                </Button>
            </Tooltip>
            <Menu
                id='simple-menu'
                anchorEl={anchorEl}
                keepMounted
                open={Boolean(anchorEl)}
                onClose={handleClose}
                getContentAnchorEl={null}
                anchorOrigin={{ vertical: "bottom", horizontal: "center" }}
                transformOrigin={{ vertical: "top", horizontal: "center" }}
                className={classes.menu}
            >
                <MenuItem value={"1"} onClick={handleClose}>
                    0.25x
                </MenuItem>
                <MenuItem value={"2"} onClick={handleClose}>
                    0.5x
                </MenuItem>
                <MenuItem value={"4"} onClick={handleClose}>
                    1x
                </MenuItem>
                <MenuItem value={"8"} onClick={handleClose}>
                    2x
                </MenuItem>
                <MenuItem value={"16"} onClick={handleClose}>
                    4x
                </MenuItem>
            </Menu>
        </div>
    );
}
