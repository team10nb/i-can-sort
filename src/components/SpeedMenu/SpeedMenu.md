### **Developer**

Ruizi Han, Shiliang Chen

Reference: MATERIAL-UI (https://material-ui.com)

### **Introduction**

Menu of speed.

It is used to adjust the speed of the animation.

### **Import**

```html
import Button from "@material-ui/core/Button"; 
import Menu from "@material-ui/core/Menu"; 
import MuiMenuItem from "@material-ui/core/MenuItem";
import { makeStyles, withStyles } from "@material-ui/core/styles"; 
import Tooltip from "@material-ui/core/Tooltip"; 
import Zoom from "@material-ui/core/Zoom";
```

### **Props**

| Name | Type | Default | Description |
| ---- | ---- | ------- | ----------- |

| handleClick | fn | | Handle clicking speed button. |
| handleClose | fn | | Handle closing speed menu. |
| speed | number | 0 | Indicate the speeding of playing animation. |
| anchorEl | object | null | Indicate the openning speed menu. |

### **Demo**

```jsx
<SpeedMenu speed={'2x'} />
```
