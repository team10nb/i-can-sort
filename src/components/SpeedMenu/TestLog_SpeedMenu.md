#### ComponentName: SpeedMenu

|Test No.|Date|Pass/Fail|Failed test(s)|Reason/Description|Modified version|
|-------|----|---------|-------------------|----------------|----------------|
|1 | 1.30 | Fail | All | No code has been constructed |version 1|
|2 | 1.30 | Pass |/|Initial version is created|version 2|
|3 | 2.4 | Fail |should match speedMenu snapshot|Change styles|version 3|
|4 | 2.4 | Pass | /                     |Snapshot updated|version 4|

##### Code - version 2
```javascript
None
```

##### Code - version 3
```javascript
import "@fontsource/roboto";
import React from "react";
import Button from "@material-ui/core/Button";
import Menu from "@material-ui/core/Menu";
import MuiMenuItem from "@material-ui/core/MenuItem";
import { makeStyles, withStyles } from "@material-ui/core/styles";
import Tooltip from "@material-ui/core/Tooltip";
import Zoom from "@material-ui/core/Zoom";

const MenuItem = withStyles({
    root: {
        justifyContent: "center",
        fontSize: "0.5em",
        fontFamily: "inherit",
    },
})(MuiMenuItem);

export default function SimpleMenu(props) {
    const { handleClick, handleClose, anchorEl, speed } = props;

    const useStyles = makeStyles({
        root: {
            display: "inline-block",
            width: "10px",
        },
        menu: {
            width: "80px",
        },
        button: {
            width: "10px",
            paddingLeft: "0px",
            paddingRight: "0px",
            fontSize: "10px",
            fontWeight: "600",
            fontFamily: "inherit",
        },
        customTooltip: {
            // I used the rgba color for the standard "secondary" color
            fontFamily: "inherit",
            fontSize: "0.5em",
            fontWeight: "700",
            paddingTop: "8px",
            paddingBottom: "10px",
            letterSpacing:"1px",
            marginTop: "3px"
        },
    });
    const classes = useStyles();

    return (
        <div className={classes.root}>
            <Tooltip
                title='Speed'
                TransitionComponent={Zoom}
                enterDelay={500}
                leaveDelay={200}
                classes={{ tooltip: classes.customTooltip }}
                arrow
            >
                <Button
                    aria-controls='simple-menu'
                    aria-haspopup='true'
                    onClick={handleClick}
                    className={classes.button}
                    style={{minWidth: "40px"}}
                >
                    {speed}
                </Button>
            </Tooltip>
            <Menu
                id='simple-menu'
                anchorEl={anchorEl}
                keepMounted
                open={Boolean(anchorEl)}
                onClose={handleClose}
                getContentAnchorEl={null}
                anchorOrigin={{ vertical: "bottom", horizontal: "center" }}
                transformOrigin={{ vertical: "top", horizontal: "center" }}
                className={classes.menu}
            >
                <MenuItem value={"1"} onClick={handleClose}>
                    0.25x
                </MenuItem>
                <MenuItem value={"2"} onClick={handleClose}>
                    0.5x
                </MenuItem>
                <MenuItem value={"4"} onClick={handleClose}>
                    1x
                </MenuItem>
                <MenuItem value={"8"} onClick={handleClose}>
                    2x
                </MenuItem>
                <MenuItem value={"16"} onClick={handleClose}>
                    4x
                </MenuItem>
            </Menu>
        </div>
    );
}


```

##### Code - version 4
```javascript
        button: {
            width: "10px",
            paddingLeft: "0px",
            paddingRight: "0px",
            fontSize: "10px",
            fontWeight: "600",
            fontFamily: "inherit",
        },
```


