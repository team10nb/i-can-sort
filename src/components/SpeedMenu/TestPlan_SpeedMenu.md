## Test Plan
--------
<br>


#### Component: SpeedMenu

|Test Name|Description|Operation|Expected Outcome|
|----|------|----------------|------------|
|should have buttons|SpeedMenu component should have a button and speed options (menuitems).|Find buttons|Buttons should be found|
|should call handleClick and close||Click SpeedMenu<br>Click a speed option|handleClick and handleClose will be called once.|
| should match SpeedMenu snapshot | Compare new snapshot with old one. | take snapshot and compare | Snapshots are the same. |

