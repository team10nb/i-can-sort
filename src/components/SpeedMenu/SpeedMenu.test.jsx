/*
    Author: Shiliang Chen, Ruizi Han
*/

import { render } from "@testing-library/react";
import SpeedMenu from "../SpeedMenu/SpeedMenu";
import userEvent from "@testing-library/user-event";


let anchorEl = null;
const handleClick = jest.fn((event)=>{anchorEl = event.currentTarget});
const handleClose = jest.fn((event)=>{anchorEl = null});
const speed = "speedMenu";
const props = {handleClick, handleClose, speed, anchorEl}

test('should have buttons', () => {
    const content = render(<SpeedMenu {...props}/>);
    const button = content.getByText("speedMenu");
    userEvent.click(button);
    content.getByText("0.25x");
    content.getByText("0.5x");
    content.getByText("1x");
    content.getByText("2x");
    content.getByText("4x");
});

test('should call handleClick and close', () => {
    const content = render(<SpeedMenu {...props}/>);
    const button = content.getByText("speedMenu");
    userEvent.click(button);
    const speedOption = content.getByText("0.25x");
    content.getByText("0.5x");
    content.getByText("1x");
    content.getByText("2x");
    content.getByText("4x");
    userEvent.click(speedOption);
    expect(handleClick).toBeCalledTimes(1);
    expect(handleClose).toBeCalledTimes(1);
})

test('should match SpeedMenu snapshot', () => {
    const content = render(<SpeedMenu {...props}/>);
    expect(content).toMatchSnapshot();
})


