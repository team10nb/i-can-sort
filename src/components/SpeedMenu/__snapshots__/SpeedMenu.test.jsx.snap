// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match SpeedMenu snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="makeStyles-root-11"
      >
        <button
          aria-controls="simple-menu"
          aria-haspopup="true"
          class="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-button-13"
          style="min-width: 40px;"
          tabindex="0"
          title="Speed"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            speedMenu
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      aria-hidden="true"
      class="MuiPopover-root makeStyles-menu-12"
      id="simple-menu"
      role="presentation"
      style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px; visibility: hidden;"
    >
      <div
        data-test="sentinelStart"
        tabindex="0"
      />
      <div
        class="MuiPaper-root MuiMenu-paper MuiPopover-paper MuiPaper-elevation8 MuiPaper-rounded"
        style="opacity: 0; transform: scale(0.75, 0.5625); visibility: hidden;"
        tabindex="-1"
      >
        <ul
          class="MuiList-root MuiMenu-list MuiList-padding"
          role="menu"
          tabindex="-1"
        >
          <li
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root WithStyles(ForwardRef(MenuItem))-root-15 MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
            role="menuitem"
            tabindex="0"
            value="1"
          >
            0.25x
            <span
              class="MuiTouchRipple-root"
            />
          </li>
          <li
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root WithStyles(ForwardRef(MenuItem))-root-15 MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
            role="menuitem"
            tabindex="-1"
            value="2"
          >
            0.5x
            <span
              class="MuiTouchRipple-root"
            />
          </li>
          <li
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root WithStyles(ForwardRef(MenuItem))-root-15 MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
            role="menuitem"
            tabindex="-1"
            value="4"
          >
            1x
            <span
              class="MuiTouchRipple-root"
            />
          </li>
          <li
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root WithStyles(ForwardRef(MenuItem))-root-15 MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
            role="menuitem"
            tabindex="-1"
            value="8"
          >
            2x
            <span
              class="MuiTouchRipple-root"
            />
          </li>
          <li
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root WithStyles(ForwardRef(MenuItem))-root-15 MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button"
            role="menuitem"
            tabindex="-1"
            value="16"
          >
            4x
            <span
              class="MuiTouchRipple-root"
            />
          </li>
        </ul>
      </div>
      <div
        data-test="sentinelEnd"
        tabindex="0"
      />
    </div>
  </body>,
  "container": <div>
    <div
      class="makeStyles-root-11"
    >
      <button
        aria-controls="simple-menu"
        aria-haspopup="true"
        class="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-button-13"
        style="min-width: 40px;"
        tabindex="0"
        title="Speed"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          speedMenu
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
